drewpc/icalparser
# SBSB-667 #comment Added a couple of checks for undefined/null values …

**Purpose**
When a client uploads a new calendar with a LOT of events, it's possible for there to be memory leaks in php when parsing those events. This creates a `ContentSourceError` and makes the calendar fail to upload, even if the calendar was already uploaded on another page (i.e. in the ticket a school had previously uploaded a calendar to their calendar page, but then was unable to upload the _same_ calendar to their event terrace). This can happen because if the number of events which causes a memory leak is Y, the calendar can be successfully uploaded at a point when it has less than Y events, and then at a later point when the calendar has Y or more events, _even if it was already uploaded elsewhere on the site_, it will fail to upload. 

**QA Steps**
1. On dev, without making any changes to your development environment, attempt to upload the following ical link as a content source, and not that it errors. 
https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics
2. You can see the same behavior on production even though the calendar was already previously uploaded to [this calendar page](https://hmhs.sau66.org/calendar). 
3. There will now be a `ContentSourceError` and an errored `ContentSource` in your database from the failed attempt at uploading. You can delete them with the postgresql cli using the following two commands (in order): 
  a. `DELETE FROM content_source_errors WHERE content_source_id IN (SELECT id FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics');`
  b. `DELETE FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics';`
  _*** If you at any point during QA run into an issue where you can't delete a `ContentSource` due to it being referenced by a `ContentItem(s)`, you can delete those `ContentItems` with the following command: `DELETE FROM content_item WHERE source_id IN (SELECT id FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics');`_
4. Now apply the `SBSB-667-ical-link-not-working` branch in the `school-blocks` repo. 
5. Imo the easiest way to get the `IcalParser` code changes into your dev environment is 
  a. In `composer.json` and `composer.lock` change the strings `"dev-schoolblocks-master"` to `"dev-SBSB-667-ical-link-not-working"`
  b. Run the following commands to update the vendor files in your `school-blocks-app-1` docker container:
    i. `docker exec -it school-blocks-app-1 sh`
    ii. `cd ../schoolblocks`
    iii. `composer install`
  _*** If you get an output about a Github token, just follow the directions to add one--it's easy to do. Note that if you copy and then paste your token, you won't see the "paste" output in Terminal, even though you DID paste it. You can just press enter and then you should be good to go._
  _******If you get an error about not being able to delete files when running `composer install`, then from outside the docker container, but inside the `school-blocks` repo run the following to remove your vendor directory: `rm -rf yii-app/protected/vendor`. Then retry the above commands.******__
7. Now stop the docker container and bring it back up again. 
8. Now try again to upload the gigantic calendar, and you should be able to get it to upload (it's going to take like a full minute, and that's with a _LOT_ of optimization--you are just literally processing 10k+ events. It's not completely impossible to optimize further, but it would require a lot of reworking... I'd really just love to move the whole thing to Watson. This should be fine for now. 
9. After it _finally_ uploads, you can run the following commands to ensure that the calendar `ContentSource` and its `ContentItems` have been added, and no `ContentSourceErrors` were created:
  a. `SELECT id, type FROM content_item WHERE source_id IN (SELECT id FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics');`
  b. `SELECT COUNT(*) FROM content_source_errors WHERE content_source_id IN (SELECT id FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics');`
  c. `SELECT id FROM content_source WHERE source_url='https://calendar.google.com/calendar/ical/sau66.org_66i576cnhh09tdm7d9jdphop58%40group.calendar.google.com/public/basic.ics';`
10. If you delete any _other_ calendar `ContentSources` from whatever page you uploaded this to, refresh the page, and look at the events terrace, you can see the events from the calendar. You should be able to see events from a up to a year in the past or a year in the future. There should not be any duplicate events. You may see recurring events, but they will have the same title but _different dates_.

**Prod Stability**
- This PR should be merged _before_ the php ticket by the same name in order to avoid things breaking.
